demo/runtime/src/safe_mix.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Substrate Demo.
+
+// Substrate Demo is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate Demo is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate Demo.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Means of mixing a series of hashes to create a single secure hash.
+//!
+//! Described in http://www.cs.huji.ac.il/~nati/PAPERS/coll_coin_fl.pdf
+
+use rstd::ops::{BitAnd, BitOr};
+use rstd::cmp;
+
+fn sub_mix<T>(seeds: &[T]) -> T where
+	T: BitAnd<Output = T> + BitOr<Output = T> + Copy
+{
+	(seeds[0] & seeds[1]) | (seeds[1] & seeds[2]) | (seeds[0] & seeds[2])
+}
+
+/// Mix a slice.
+pub fn mix<T>(seeds: &[T]) -> Result<T, ()> where
+	T: BitAnd<Output = T> + BitOr<Output = T>,
+	T: Default + Copy
+{
+	Ok(seeds.iter().cloned().mixed())
+}
+
+/// The mixed trait for mixing a ssequence.
```

Comment:
typo. and why use a trait if it's supposed to be an implementation of a single algorithm? why not `fn mix<I, T>` instead of `impl<I, T> Mixed for I`?

Reply:
why not? is this not idiomatic?